File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nextLargerNodes(self, head: Optional[ListNode]) -> List[int]:
        stack, ans = [], []
        while head:
            while stack and stack[-1][0] < head.val:
                ans[stack.pop()[1]] = head.val
            stack.append([head.val, len(ans)])
            ans.append(0)
            head = head.next
        return ans

File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number[]}
 */
var nextLargerNodes = function(head) {
  const stack = [];
  const ans = [];
  
  while (head) {
    while (stack.length && stack.at(-1)[0] < head.val) {
      const index = stack.pop()[1];
      ans[index] = head.val;
    }
    
    stack.push([head.val, ans.length]);
    ans.push(0);
    head = head.next;
  }
  
  return ans;
};